internal/docs/design.md

```diff
@@ -184,6 +190,158 @@ needed). However, this would result in inconsistent names over time (e.g., some
+  // ...
+}
+
+// OpenWidget calls OpenWidgetURL with the URL parsed from urlstr.
+// OpenWidget is safe to call from multiple goroutines.
+func (mux *URLMux) OpenWidget(ctx context.Context, urlstr string) (*Widget, error) {
+  u, err := url.Parse(urlstr)
+  if err != nil {
+    return nil, fmt.Errorf("open widget: %v", err)
+  }
+  return mux.OpenWidgetURL(ctx, u)
+}
+
+// OpenWidgetURL dispatches the URL to the opener that is registered with the
+// URL's scheme. OpenWidget is safe to call from multiple goroutines.
+func (mux *URLMux) OpenWidgetURL(ctx context.Context, u *url.URL) (*Widget, error) {
+  // ...
+}
+
+// DefaultURLMux returns the URLMux used by OpenWidget.
+func DefaultURLMux() *URLMux {
+  return defaultURLMux
+}
+
+var defaultURLMux = new(URLMux)
+
+// OpenWidget opens the Widget identified by the URL given. URL openers must be
+// registered in the DefaultURLMux, which is typically done in driver
+// packages' initialization.
+func OpenWidget(ctx context.Context, urlstr string) (*Widget, error) {
+  return DefaultURLMux().OpenWidget(urlstr)
+}
+```
+
+The repetition of `Widget` in the method names permits a type to handle multiple
+resources within the API. Exporting the `URLMux` allows applications to build
+their own muxes, potentially wrapping existing.
+
+Driver packages should include their own `URLOpener` struct type which
+implements all the relevant `WidgetURLOpener` methods. The URL should only serve
+to identify which resource to open. Any credentials or other complex values
+should be taken in as struct fields, not as input from URL. If the driver
+package registers its `URLOpener` with the `DefaultURLMux`, then it should
+populate these complex fields from environment variables. If doing so is
+undesirable or expensive, then it should not register with the `DefaultURLMux`
+and instead rely on users to create their own mux. If there already exists a
+well-established URI format for the backend (like S3 URLs or database connection
+URIs), then drivers should honor them where possible.
+
+[Backing services factor]: https://12factor.net/backing-services
+
+#### URL Examples
+
+A `WidgetURLOpener` implementation for a hypothetical GCP service:
+
+```go
+package gcpfoo
+
+// ...
+
+const Scheme = "gcpwidget"
+
+type URLOpener struct {
+  Client  *gcp.HTTPClient
+  Options Options
+}
+
+func (o *URLOpener) OpenWidgetURL(ctx context.Context, u *url.URL) (*foo.Widget, error) {
+  // ...
+  return OpenWidget(ctx, o.Client, u.Host, &o.Options)
+}
+
+type lazyURLOpener struct {
+  init sync.Once
+  o    *URLOpener
+  err  error
+}
+
+func (o *lazyURLOpener) OpenWidgetURL(ctx context.Context, u *url.URL) (*foo.Widget, error) {
+  o.init.Once(func() {
+    creds, err := gcp.DefaultCredentials(ctx)
+    if err != nil {
+      o.err = err
+      return
+    }
+    o.o = new(URLOpener)
+    o.o.Client, _ = gcp.NewHTTPClient(http.DefaultTransport, creds.TokenSource)
+  })
+  return o.OpenWidgetURL(ctx, u)
```

Comment:
I think you meant `o.o.OpenWidgetURL` here, but see above.

Reply:
Fixed with better names.